msticpy/context/contextproviders/context_provider_base.py

```diff
@@ -0,0 +1,297 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+Module for ContextProvider classes.
+
+Input can be a single observable or a pandas DataFrame containing
+    multiple observables. Processing may require a an API key and
+processing performance may be limited to a specific number of
+requests per minute for the account type that you have.
+
+"""
+from abc import abstractmethod
+from typing import Tuple, Any, List, Union, Dict, Iterable
+from functools import lru_cache
+
+import pandas as pd
+
+from ..provider_base import Provider
+from ...common.utility import export
+from ..._version import VERSION
+
+__version__ = VERSION
+__author__ = "Ian Hellen"
+
+
+@export
+class ContextProvider(Provider):
+    """Abstract base class for Context providers."""
+
+    _REQUIRED_PARAMS: List[str] = []
+
+    @lru_cache(maxsize=256)
+    def lookup_item(
+        self, item: str, item_type: str = None, query_type: str = None, **kwargs
+    ) -> pd.DataFrame:
+        """
+        Lookup from a value.
+
+        Parameters
+        ----------
+        item : str
+            item to lookup
+        item_type : str, optional
+            The Type of the item to lookup, by default None (type will be inferred)
+        query_type : str, optional
+            Specify the data subtype to be queried, by default None.
+            If not specified the default record type for the item_value
+            will be returned.
+
+        Returns
+        -------
+        pd.DataFrame
+            The lookup result:
+            result - Positive/Negative,
+            details - Lookup Details (or status if failure),
+            raw_result - Raw Response
+            reference - URL of the item
+
+        Raises
+        ------
+        NotImplementedError
+            If attempting to use an HTTP method or authentication
+            protocol that is not supported.
+
+        Notes
+        -----
+        Note: this method uses memoization (lru_cache) to cache results
+        for a particular observable to try avoid repeated network calls for
+        the same item.
+
+        """
+        return self.lookup_observables(
```

Comment:
I'm not yet convinced this is working properly yet. I probably need to rewrite a bit lookup_observables/lookup_iocs not to call lookup_item/lookup_ioc or it will lead to a loop of dependent functions

Reply:
DId you verify that this works OK?

Reply:
I applied some changes, and I confirm it now works as expected.
The tests are in https://github.com/microsoft/msticpy/pull/556